### PR TITLE
fix: fix azcopy exec timeout func

### PR DESCRIPTION
**Pull request by umagnus**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

fix: fix azcopy exec timeout func, use WaitForExecCompletion func in util.go waits for the exec function to complete or times out

**Which issue(s) this PR fixes**:

Fixes #

**Requirements**:
- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:
```
none
```

### CODE REVIEW

File: pkg/util/util.go [hunk 1]
```diff
@@ -387,3 +388,30 @@ func SetVolumeOwnership(path, gid, policy string) error {
...
 }
+
+// ExecFunc returns a exec function's output and error
+type ExecFunc func() (err error)
+
+// TimeoutFunc returns output and error if an ExecFunc timeout
+type TimeoutFunc func() (err error)
+
+// WaitForExecCompletion waits for the exec function to complete or times out
+func WaitForExecCompletion(timeout time.Duration, execFunc ExecFunc, timeoutFunc TimeoutFunc) error {
```
Comment (andyzhangx) [inline pkg/util/util.go:399]: pls add a ut for this func

Comment (umagnus) [reply pkg/util/util.go:399]: done

File: pkg/blob/controllerserver.go [hunk 2]
```diff
@@ -759,43 +757,39 @@ func (d *Driver) copyBlobContainer(req *csi.CreateVolumeRequest, accountSasToken
...
-					klog.Warningf("CopyBlobContainer(%s, %s, %s) failed with error(%v): %v", resourceGroupName, accountName, dstPath, copyErr, string(out))
-				} else {
-					klog.V(2).Infof("copied blob container %s to %s successfully", srcContainerName, dstContainerName)
-				}
-				return copyErr
+	case util.AzcopyJobRunning:
+		return fmt.Errorf("an existed azcopy job is running, copy percent: %s%%, please wait for it to complete", percent)
+	case util.AzcopyJobNotFound:
+		klog.V(2).Infof("copy blob container %s to %s", srcContainerName, dstContainerName)
+		copyErr := util.WaitForExecCompletion(time.Duration(d.waitForAzCopyTimeoutMinutes)*time.Minute, func() error {
```
Comment (andyzhangx) [inline pkg/blob/controllerserver.go:772]: this is hard to read, can you define execFunc, timeoutFunc as a var outside?

File: pkg/util/util.go [hunk 3]
```diff
@@ -387,3 +388,30 @@ func SetVolumeOwnership(path, gid, policy string) error {
...
 	return volume.SetVolumeOwnership(&VolumeMounter{path: path}, path, &gidInt64, &fsGroupChangePolicy, nil)
 }
+
+// ExecFunc returns a exec function's output and error
+type ExecFunc func() (err error)
+
+// TimeoutFunc returns output and error if an ExecFunc timeout
+type TimeoutFunc func() (err error)
+
+// WaitForExecCompletion waits for the exec function to complete or times out
```
Comment (andyzhangx) [inline pkg/util/util.go:398]: WaitForExecCompletion waits for the exec function to complete or return timeout error

File: pkg/blob/controllerserver.go [hunk 4]
```diff
@@ -759,43 +757,39 @@ func (d *Driver) copyBlobContainer(req *csi.CreateVolumeRequest, accountSasToken
...
-				}
-				out, copyErr := cmd.CombinedOutput()
-				if copyErr != nil {
-					klog.Warningf("CopyBlobContainer(%s, %s, %s) failed with error(%v): %v", resourceGroupName, accountName, dstPath, copyErr, string(out))
-				} else {
-					klog.V(2).Infof("copied blob container %s to %s successfully", srcContainerName, dstContainerName)
-				}
-				return copyErr
+	case util.AzcopyJobRunning:
+		return fmt.Errorf("an existed azcopy job is running, copy percent: %s%%, please wait for it to complete", percent)
```
[the remark this answers is not in the document]
Comment (umagnus) [reply pkg/blob/controllerserver.go]: done

File: pkg/blob/controllerserver.go [hunk 5]
```diff
@@ -759,43 +757,39 @@ func (d *Driver) copyBlobContainer(req *csi.CreateVolumeRequest, accountSasToken
...
 			}
-		case <-timeAfter:
-			return fmt.Errorf("timeout waiting for copy blob container %s to %s succeed", srcContainerName, dstContainerName)
+			if out, err := cmd.CombinedOutput(); err != nil {
+				return fmt.Errorf("exec error: %v, output: %v", err, string(out))
+			}
+			return nil
+		}, func() error {
+			_, percent, _ := d.azcopy.GetAzcopyJob(dstContainerName, authAzcopyEnv)
+			return fmt.Errorf("timeout waiting for copy blob container %s to %s succeed, copy percent: %s%%", srcContainerName, dstContainerName, percent)
```
[the remark this answers is not in the document]
Comment (umagnus) [reply pkg/blob/controllerserver.go]: done

File: pkg/blob/controllerserver.go (hunk 2 above)
Comment (umagnus) [reply pkg/blob/controllerserver.go]: done

File: pkg/util/util.go (hunk 3 above)
Comment (umagnus) [reply pkg/util/util.go]: done

**Review by andyzhangx — COMMENTED** [review]
/ok-to-test

**Review by andyzhangx — APPROVED** [review]
/lgtm

**Pull request merged**